core/src/main/java/org/infinispan/persistence/support/ComposedSegmentedLoadWriteStore.java

```diff
@@ -191,15 +193,21 @@ public void deleteBatch(Iterable<Object> keys) {
    }
 
    @Override
-   public void writeBatch(Iterable<MarshalledEntry<? extends K, ? extends V>> marshalledEntries) {
-      Flowable.fromIterable(marshalledEntries)
-            // Separate out batches by segment
+   public CompletionStage<Void> writeBatch(Publisher<MarshallableEntry<? extends K, ? extends V>> publisher) {
+      CompletableFuture<Void> future = new CompletableFuture<>();
+
+      Flowable.fromPublisher(publisher)
             .groupBy(me -> keyPartitioner.getSegment(me.getKey()))
-            .blockingForEach(groupedFlowable ->
+            .serialize()
+            .doOnNext(groupedFlowable ->
```

Comment:
You should use `flatMap` here instead. To be honest we could parallelize this as I mentioned above, but lets just keep it simple for now.